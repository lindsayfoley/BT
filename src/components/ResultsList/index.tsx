import { FunctionComponent } from 'react';
import ResultsCard, { IArticle } from '../ResultsCard';

export interface IResultsListProps {
  list: IArticle[];
}

const ResultsList: FunctionComponent<IResultsListProps> = ({ list }) => {
  if (!list || list.length === 0) {
    return null;
  }

  return (
    <section>
      <h2>{`Showing ${list.length} results:`}</h2>
      {list.map((article, index) => (
        <ResultsCard {...article} key={`${article.title}-${index}`} />
      ))}
    </section>
  );
};

export default ResultsList;
