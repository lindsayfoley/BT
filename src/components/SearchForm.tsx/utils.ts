import { IArticle } from "../ResultsCard";

export const NEWS_API_URL = 'https://newsapi.org/v2/';
export const MAX_RESULTS = 10;

export interface IResponseDataProps {
    status: string;
    totalResults: number;
    articles: IArticle[];
  }
  
  export enum EErrorMessage {
    Generic = 'Sorry an error has occurred',
    NoResults = 'Sorry no results were found',
    Invalid = 'Please enter a valid search term',
    Illegal = 'The following characters are not allowed: <, >, ), (',
    RejectedPromise = 'Sorry, something went wrong, please try again later',
  }
  