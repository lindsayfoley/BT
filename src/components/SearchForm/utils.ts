import { IArticle } from '../ResultsCard';
export interface IResponseDataProps {
  status: string;
  totalResults: number;
  articles: IArticle[];
}

export enum EErrorMessage {
  Generic = 'Sorry an error has occurred',
  NoResults = 'Sorry no results were found',
  Invalid = 'Please enter a valid search term',
  Illegal = 'The following characters are not allowed: <, >, ), (',
  RejectedPromise = 'Sorry, something went wrong, please try again later',
}