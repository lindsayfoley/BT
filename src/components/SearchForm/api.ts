import { NEWS_API_URL, MAX_RESULTS } from "./utils";

export const getArticles = async (article: string) => {
    const response = await fetch(
      `${NEWS_API_URL}everything?q=${article}&apiKey=${process.env.REACT_APP_NEWS_API_KEY}&pageSize=${MAX_RESULTS}`
    );
    return await response.json();
  };
  