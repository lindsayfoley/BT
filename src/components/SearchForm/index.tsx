import React, {
  ChangeEvent,
  FormEvent,
  FunctionComponent,
  useState,
} from 'react';
import { IArticle } from '../ResultsCard';
import ResultsList from '../ResultsList';
import { getArticles } from './api';
import {
  EErrorMessage,
  IResponseDataProps
} from './utils';

const SearchForm: FunctionComponent = () => {
  const [searchTerm, setSearchTerm] = useState<string>('');
  const [articleResults, setArticleResults] = useState<IArticle[] | null>(null);
  const [errorMessage, setErrorMessage] = useState<string | null>(null);

  const validateData = (data: IResponseDataProps) => {
    const { status, totalResults } = data;

    if (status !== 'ok') {
      setErrorMessage(EErrorMessage.Generic);
    } else if (totalResults === 0) {
      setErrorMessage(EErrorMessage.NoResults);
      setArticleResults(null);
    } else {
      setArticleResults(data.articles);
    }
  };

  const validateSearchTerm = (term: string) => {
    const validCharacters = new RegExp('[a-zA-Z0-9]');
    const illegalCharacters = new RegExp('[<|>|(|)]');
    const termIsInvalid = term.replace(/\s/g, '').length === 0;

    const hasIllegalCharacters = illegalCharacters.test(term);
    const hasOnlyInvalidCharacters = !validCharacters.test(term);
    const isInvalidFirstCharacter =
      term.length === 1 && !validCharacters.test(term);

    if (hasIllegalCharacters) {
      setErrorMessage(EErrorMessage.Illegal);
      setArticleResults(null);
      return;
    }

    if (termIsInvalid || isInvalidFirstCharacter || hasOnlyInvalidCharacters) {
      setErrorMessage(EErrorMessage.Invalid);
      setArticleResults(null);
    } else {
      setErrorMessage(null);
    }
  };

  const handleOnChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
    validateSearchTerm(event.target.value);
  };

  const handleSubmit = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    if (errorMessage) {
      return;
    }

    const term = encodeURIComponent(searchTerm.trim());

    try {
      const data = await getArticles(term);
      validateData(data);
    } catch (e) {
      setErrorMessage(EErrorMessage.RejectedPromise);
    }
  };

  return (
    <main>
      <h1>BT React Code Test - by Lindsay Foley - 28/1/21</h1>
      <hr />
      {errorMessage && <h3>{errorMessage}</h3>}
      <form onSubmit={handleSubmit}>
        <label htmlFor="article-searchbox">Start your search for a news article</label>
        <input
          id="article-searchbox"
          type="text"
          value={searchTerm}
          placeholder="Enter article"
          onChange={handleOnChange}
          required
        />
        <input type="submit" value="Submit search" />
      </form>
      {articleResults && !errorMessage && <ResultsList list={articleResults} />}
    </main>
  );
};

export default SearchForm;
