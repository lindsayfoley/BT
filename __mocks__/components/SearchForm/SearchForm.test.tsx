import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import SearchForm from '../../../src/components/SearchForm';
import { getArticles } from '../../../src/components/SearchForm/api';
import { mockArticleList } from '../ResultsList/ResultsList.test';
import ResultsList from '../../../src/components/ResultsList';
import { act } from 'react-dom/test-utils';
import { EErrorMessage } from '../../../src/components/SearchForm/utils';

jest.mock('../../../src/components/SearchForm/api');
const mockGetArticles = getArticles as jest.MockedFunction<typeof getArticles>;
const searchTerm = 'Pizza';
const mockEvent = {
  target: { value: searchTerm },
};
const flushPromises = () => new Promise(setImmediate);

const setup = () => shallow(<SearchForm />);

describe('<SearchForm />', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = setup();
  });

  it('renders without crashing', () => {
    expect(wrapper.isEmptyRender()).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a form with all required fields', () => {
    expect(wrapper.find('label')).toHaveLength(1);
    expect(wrapper.find('#article-searchbox')).toHaveLength(1);
    expect(wrapper.find('input[type="submit"]')).toHaveLength(1);
  });

  it('should update the text in the input field to match the term entered', () => {
    expect(wrapper.find('#article-searchbox').prop('value')).toContain('');
    wrapper.find('#article-searchbox').simulate('change', mockEvent);
    expect(wrapper.find('#article-searchbox').prop('value')).toBe(searchTerm);
  });

  describe('when the search term is valid', () => {
    it('it should call the function to get data with the search term entered', () => {
      wrapper.find('#article-searchbox').simulate('change', mockEvent);
      wrapper.find('form').simulate('submit', { preventDefault: () => {} });
      expect(mockGetArticles).toHaveBeenCalledWith(searchTerm);
    });

    it('should render search results when the promise to fetch is resolved', async () => {
      expect(wrapper.find(ResultsList).exists()).toBeFalsy();
      act(() => {
        wrapper.find('#article-searchbox').simulate('change', mockEvent);
        mockGetArticles.mockResolvedValue({
          articles: mockArticleList,
          status: 'ok',
        });
        wrapper.find('form').simulate('submit', { preventDefault: () => {} });
      });
      await flushPromises();
      wrapper.update();
      expect(wrapper.find(ResultsList).exists()).toBeTruthy();
    });
  });

  describe('when the search is unsuccessful', () => {
    it('should display an error message if the promise rejects', async () => {
      act(() => {
        wrapper.find('#article-searchbox').simulate('change', mockEvent);
        mockGetArticles.mockRejectedValue(new Error('Async error'));
        wrapper.find('form').simulate('submit', { preventDefault: () => {} });
      });
      await flushPromises();
      wrapper.update();
      expect(wrapper.find('h3').text()).toContain(
        EErrorMessage.RejectedPromise
      );
      expect(wrapper.find(ResultsList).exists()).toBeFalsy();
    });

    it('should display an error message if no results are returned for the search term', async () => {
      act(() => {
        wrapper.find('#article-searchbox').simulate('change', mockEvent);
        mockGetArticles.mockResolvedValue({
          articles: [],
          status: 'ok',
          totalResults: 0,
        });
        wrapper.find('form').simulate('submit', { preventDefault: () => {} });
      });
      await flushPromises();
      wrapper.update();
      expect(wrapper.find('h3').text()).toContain(EErrorMessage.NoResults);
    });

    it('should display an generic error message if the the returned status is not "ok"', async () => {
      act(() => {
        wrapper.find('#article-searchbox').simulate('change', mockEvent);
        mockGetArticles.mockResolvedValue({
          articles: [],
          status: 'bad',
          totalResults: 0,
        });
        wrapper.find('form').simulate('submit', { preventDefault: () => {} });
      });
      await flushPromises();
      wrapper.update();
      expect(wrapper.find('h3').text()).toContain(EErrorMessage.Generic);
    });

    it('should display an error message if the term typed into the search box solely contains special characters', () => {
      wrapper.find('#article-searchbox').simulate('change', {
        target: { value: '%$Â£@!' },
      });
      wrapper.find('form').simulate('submit', { preventDefault: () => {} });
      expect(wrapper.find('h3').text()).toContain(EErrorMessage.Invalid);
    });

    it('should display an error message if the term typed into the search box is a single special character', () => {
      wrapper.find('#article-searchbox').simulate('change', {
        target: { value: '*' },
      });
      wrapper.find('form').simulate('submit', { preventDefault: () => {} });
      expect(wrapper.find('h3').text()).toContain(EErrorMessage.Invalid);
    });

    it('should display an error message if the term typed into the search box contains illegal characters', () => {
      wrapper.find('#article-searchbox').simulate('change', {
        target: { value: '(' },
      });
      wrapper.find('form').simulate('submit', { preventDefault: () => {} });
      expect(wrapper.find('h3').text()).toContain(EErrorMessage.Illegal);
    });
  });
});
