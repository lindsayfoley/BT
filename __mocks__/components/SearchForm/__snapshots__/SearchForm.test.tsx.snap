// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchForm /> renders without crashing 1`] = `
<main>
  <h1>
    BT React Code Test - by Lindsay Foley - 28/1/21
  </h1>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="article-searchbox"
    >
      Start your search for a news article
    </label>
    <input
      id="article-searchbox"
      onChange={[Function]}
      placeholder="Enter article"
      required={true}
      type="text"
      value=""
    />
    <input
      type="submit"
      value="Submit search"
    />
  </form>
</main>
`;
